### DISCUSSION TITLE
Shutting down all background Isolates and starting a new one

### DISCUSSION THREAD

**Discussion by JerContact** [simolus3/drift#2125] [Q&A]
I'm having touble figuring out the best way to shut down all DriftIsolates and then restarting a new one right after it.  For my app, I need to do this action a lot.

Currently I'm using this in the provider to setup the background isolate since I need to use path, this is straight from the documentation example, just curious if there is a best practice to get the background driftisolate, shut them all down, and then re-connect a new one.

`void main() async {
runApp(Provider<MyDatabase>(
create: (context) => MyDatabase.connect(createdDriftIsolateAndConnect()),
child: MyApp(),
dispose: (context, db) => db.close,
)
}`

**Comment by simolus3** [comment]
> For my app, I need to do this action a lot.

May I ask why?

> just curious if there is a best practice to get the background driftisolate

Drift doesn't have a central list of active `DriftIsolate`s anywhere. You'd have to manually dispose them by calling `shutdownAll`.

I'm thinking about adding a parameter like `singleClient` to `DriftIsolate` which will only allow a single client to connect and the isolate to automatically close itself when that client disconnects. But I'm curious to hear about your use-case first to see if that would work here.

**Reply by JerContact** [reply]
I wrote this earlier, and then sent that issue up on another thread up here, but my issue is this with my app is a type as you go search.  I would like to kill all isolates and startup another one so that only one isolate is working at a time.  Sort of the opposite of what you are saying, have an override isolate call or something like that.

**Reply by simolus3** [reply]
> I would like to kill all isolates and startup another one so that only one isolate is working at a time.

Perhaps there's a misunderstanding here. Starting a new isolate _for every database query_ is much worse for performance than just keeping a single long-running `DriftIsolate` alive all the time. In recent Dart and Flutter versions, idle isolates don't take up many resources. What a drift isolate is doing is essentially:

1. Wait for a client (your database instance living on the UI thread) for send a query.
2. Execute that query and send the result back to the client.
3. Goto 1.

So in reality, a drift isolate will spend most of its time waiting for a client to send a query, and it consumes virtually no resources while waiting. I recommend keeping a single instance of your database class alive during your app's lifecycle.

**Reply by JerContact** [reply]
This isn't acceptable as sometimes the typing is faster than the isolate getting back.  That's why I need to close it/cancel it and start up a new isolate.  In a type-ahead search, this slows down the search tremendously to have to wait in sequence for every letter, if that makes sense
